## Homework 2: Similarity

One of the most important metrics that we have in Data Science is the ability to group similar data together.  In Lecture 1, you learned about several metrics of similarity; in the past few days, you filled out a survey to share some bits about you.

### Puzzle 1: Import pandas + read_csv

The dataset `hello-world-cleaned.csv` is provided for you in same directory as this Notebook.  Use `pandas` to read this file:

In [4]:
import pandas as pd


df = pd.read_csv('hello-world-cleaned.csv')

#Set number of columns and rows
numCols = len(df.columns)
categories = df.columns
numRows = len(df.index)

print(df)

   What's your first name?  \
0                     Wade   
1                   Aashna   
2                 Abdullah   
3                   Aganze   
4                       AJ   
5                       AJ   
6                     Ajay   
7                   Akarsh   
8                   Albert   
9                     Alok   
10                  Ananya   
11                  Andrew   
12                   Asher   
13                 Atharva   
14                  Austin   
15                 Avinash   
16                  Avinda   
17                    Bing   
18                   Brian   
19               Catherine   
20                  Charan   
21                   Cheng   
22                   Chris   
23               Christian   
24               Christine   
25                   Daksh   
26                   David   
27                   David   
28                   David   
29                   Derek   
..                     ...   
59                 Matthew   
60        

### Puzzle 2: Find yourself

Use a selector to find **yourself** within the DataFrame.  It may be useful to store yourself in a variable called `my_row` or similar:

In [5]:
mybool = (df["What's your first name?"] == "Chad")
my_row = df[mybool]

print(my_row)




   What's your first name?  \
87                    Chad   

   What year in college do you consider yourself (ex: Senior)?  \
87                                           Graduate            

   What major do you consider yourself (ex: CS+STAT)?  \
87                                            Physics   

   Cell Phone: Android or iOS? Tabs or Spaces? Favorite Code Editor?  \
87                         iOS            Tabs                  Atom   

   Favorite Programming Language? Favorite Candy?  \
87                         Python        Licorice   

   Favorite IDE Theme (ex: dark/light) Spirit Animal?  ...  \
87                                Dark         dragon  ...   

        Favorite Food?  Favorite Color?  Marvel or DC?  \
87  Mediterranean food              red             dc   

   Instrument you play (if any)?  \
87                          none   

   The Dress: https://en.wikipedia.org/wiki/The_dress#/media/File:The_Dress_(viral_phenomenon).png  \
87                    

### Puzzle 3: Add a similarity metric to every person

Add a column to the DataFrame that ranks how similar each person is to you.  A correct algorithm should have your similarity  score with yourself being of maximum value (and the similarity score of everyone else to be slightly lower).

In [6]:

#fills in the NaNs to avoid errors

#>>> values = {'A': 0, 'B': 1, 'C': 2, 'D': 3}
# df.fillna(value=values)
#    A   B   C   D
#0   0.0 2.0 2.0 0
#1   3.0 4.0 2.0 1
#2   0.0 1.0 2.0 5
#3   0.0 3.0 2.0 4 

for i in range(numCols):
   
    

        
  if (type(df[categories[i]][87]) == str ):
       print("is string", df[categories[i]].dtypes)
       df[categories[i]] = df[categories[i]].fillna("Neither")

  else:  
        df[categories[i]] = df[categories[i]].fillna(9999)
        



#print(my_row.iloc[ 0][colNum-1])
#print(my_row.iloc[ 0][0] == df[categories[0]][87])

#Initializes New column
df["Similarity"] =   df[categories[0]].astype(str).str.find(my_row.iloc[0][0]) 

#Produces a  -1 for every difference between me and the other person 
for i in range(1,numCols):
    #print("is this true",is_string_dtype(df[categories[i]][1]))
    #print(df[categories[i]][64])
    
    if type(df[categories[i]]) == str:
       df["Similarity"] =  df["Similarity"] + df[categories[i]].str.find(my_row.iloc[0][i]) 
      # print("does it come in here")

    else:
      # print("in other loop")
       #print(i)
       for j in range(numRows):
           if(my_row.iloc[0][i] != df[categories[i]][j]): 
              df["Similarity"][j] = df["Similarity"][j] - 1
              
        
              
       

df["Similarity"] = (df["Similarity"] + numCols)/ numCols
print(numCols)
print()
               
             
print(df)
#df.loc[df["What's your first name?"] == "Chad"]


is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object
is string object


C:\Users\DELL\Anaconda3\lib\site-packages\ipykernel_launcher.py:47: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


23

   What's your first name?  \
0                     Wade   
1                   Aashna   
2                 Abdullah   
3                   Aganze   
4                       AJ   
5                       AJ   
6                     Ajay   
7                   Akarsh   
8                   Albert   
9                     Alok   
10                  Ananya   
11                  Andrew   
12                   Asher   
13                 Atharva   
14                  Austin   
15                 Avinash   
16                  Avinda   
17                    Bing   
18                   Brian   
19               Catherine   
20                  Charan   
21                   Cheng   
22                   Chris   
23               Christian   
24               Christine   
25                   Daksh   
26                   David   
27                   David   
28                   David   
29                   Derek   
..                     ...   
59                 Matthew   
60    

#### Validation 1: Are you most similar to yourself?

In [7]:
# Should print out your own row (feel free to change `df` or 'Similarity' to a different variable/column name):
df.nlargest(1, 'Similarity')

What's your first name?  \
87                    Chad   

   What year in college do you consider yourself (ex: Senior)?  \
87                                           Graduate            

   What major do you consider yourself (ex: CS+STAT)?  \
87                                            Physics   

   Cell Phone: Android or iOS? Tabs or Spaces? Favorite Code Editor?  \
87                         iOS            Tabs                  Atom   

   Favorite Programming Language? Favorite Candy?  \
87                         Python        Licorice   

   Favorite IDE Theme (ex: dark/light) Spirit Animal?  ...  Favorite Color?  \
87                                Dark         dragon  ...              red   

    Marvel or DC?  Instrument you play (if any)?  \
87             dc                           none   

   The Dress: https://en.wikipedia.org/wiki/The_dress#/media/File:The_Dress_(viral_phenomenon).png  \
87                                    Yellow and Gold                                                

   How many credit hours are you in this semester?  \
87                                            13.0   

   Excluding CS 225 and CS 296, favorite course at UIUC so far?  (This semester or a previous semester.)  \
87                                        Physics 504                                                      

   Eating Out: # of meals out /week? Height (in)  Windows, Mac, or Linux?  \
87                               0.0        72.0                  Windows   

   Similarity  
87        1.0  

[1 rows x 24 columns]

### Final Result: Who is most similar to you?

Excluding yourself, who in CS 296 is/are most similar to you?

In [8]:
df.nlargest(4, 'Similarity')

What's your first name?  \
87                    Chad   
5                       AJ   
68                 Rolando   
86               Zhengguan   

   What year in college do you consider yourself (ex: Senior)?  \
87                                           Graduate            
5                                           Sophomore            
68                                             Junior            
86                                          Sophomore            

   What major do you consider yourself (ex: CS+STAT)?  \
87                                            Physics   
5                                          CS + Stats   
68                                                 CS   
86                                          Aerospace   

   Cell Phone: Android or iOS? Tabs or Spaces? Favorite Code Editor?  \
87                         iOS            Tabs                  Atom   
5                          iOS            Tabs                  Atom   
68                         iOS            Tabs                  Atom   
86                         iOS            Tabs                  Atom   

   Favorite Programming Language? Favorite Candy?  \
87                         Python        Licorice   
5                            Java       Milky Way   
68                         Python    Gummy Snacks   
86                         Python         Hi-chew   

   Favorite IDE Theme (ex: dark/light) Spirit Animal?  ...  Favorite Color?  \
87                                Dark         dragon  ...              red   
5                                 Dark           Wolf  ...            Black   
68                                Dark          Llama  ...            Black   
86                                Dark           Crow  ...             Blue   

    Marvel or DC?  Instrument you play (if any)?  \
87             dc                           none   
5          Marval                          Piano   
68         Marval                         Guitar   
86         Marval                      Harmonica   

   The Dress: https://en.wikipedia.org/wiki/The_dress#/media/File:The_Dress_(viral_phenomenon).png  \
87                                    Yellow and Gold                                                
5                                     Yellow and Gold                                                
68                                    Yellow and Gold                                                
86                                     Blue and Black                                                

   How many credit hours are you in this semester?  \
87                                            13.0   
5                                             18.0   
68                                            15.0   
86                                            15.0   

   Excluding CS 225 and CS 296, favorite course at UIUC so far?  (This semester or a previous semester.)  \
87                                        Physics 504                                                      
5                                            STAT 400                                                      
68                                             CS 361                                                      
86                                              AE199                                                      

   Eating Out: # of meals out /week? Height (in)  Windows, Mac, or Linux?  \
87                               0.0        72.0                  Windows   
5                                0.0        68.0                  Windows   
68                               2.0        69.0                  Windows   
86                              10.0        72.0                      Max   

   Similarity  
87   1.000000  
5    0.304348  
68   0.304348  
86   0.304348  

[4 rows x 24 columns]